mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/PostgresMessageMapper.java

```diff
@@ -0,0 +1,373 @@
+            .flatMap(totalMessage -> mailboxMessageDAO.countUnseenMessagesByMailboxId((PostgresMailboxId) mailbox.getMailboxId())
+                .map(unseenMessage -> MailboxCounters.builder()
+                    .mailboxId(mailbox.getMailboxId())
+                    .count(totalMessage)
+                    .unseen(unseenMessage)
+                    .build()));
+    }
+
+    @Override
+    public void delete(Mailbox mailbox, MailboxMessage message) throws MailboxException {
+        deleteMessages(mailbox, List.of(message.getUid()));
+    }
+
+    @Override
+    public Map<MessageUid, MessageMetaData> deleteMessages(Mailbox mailbox, List<MessageUid> uids) {
+        return deleteMessagesReactive(mailbox, uids).block();
+    }
+
+    @Override
+    public Mono<Map<MessageUid, MessageMetaData>> deleteMessagesReactive(Mailbox mailbox, List<MessageUid> uids) {
+        return mailboxMessageDAO.findMessagesByMailboxIdAndUIDs((PostgresMailboxId) mailbox.getMailboxId(), uids)
+            .collectMap(MailboxMessage::getUid, MailboxMessage::metaData)
+            .flatMap(map -> mailboxMessageDAO.deleteByMailboxIdAndMessageUids((PostgresMailboxId) mailbox.getMailboxId(), uids)
+                .then(Mono.just(map)));
+    }
+
+    @Override
+    public MessageUid findFirstUnseenMessageUid(Mailbox mailbox) {
+        return mailboxMessageDAO.findFirstUnseenMessageUid((PostgresMailboxId) mailbox.getMailboxId()).block();
+    }
+
+    @Override
+    public Mono<Optional<MessageUid>> findFirstUnseenMessageUidReactive(Mailbox mailbox) {
+        return mailboxMessageDAO.findFirstUnseenMessageUid((PostgresMailboxId) mailbox.getMailboxId())
+            .map(Optional::of)
+            .switchIfEmpty(Mono.just(Optional.empty()));
+    }
+
+    @Override
+    public List<MessageUid> findRecentMessageUidsInMailbox(Mailbox mailbox) {
+        return findRecentMessageUidsInMailboxReactive(mailbox).block();
+    }
+
+    @Override
+    public Mono<List<MessageUid>> findRecentMessageUidsInMailboxReactive(Mailbox mailbox) {
+        return mailboxMessageDAO.findAllRecentMessageUid((PostgresMailboxId) mailbox.getMailboxId())
+            .collectList();
+    }
+
+    @Override
+    public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws MailboxException {
+        return addReactive(mailbox, message).block();
+    }
+
+    @Override
+    public Mono<MessageMetaData> addReactive(Mailbox mailbox, MailboxMessage message) {
+        return Mono.fromCallable(() -> {
+                message.setSaveDate(Date.from(clock.instant()));
+                return message;
+            })
+            .flatMap(mailboxMessage -> setNewMessageUid(mailboxMessage)
+                .then(setNewMessageModSeq(mailboxMessage)))
+            .then(mailboxMessageDAO.insert(message))
+            .then(saveBodyContent(message)
+                .flatMap(blobId -> messageDAO.insert(message, blobId.asString())))
```

Comment:
The order of execution MUST be: `saveBodyContent` then `insert messageDAO` then `insert mailboxMessageDAO`.

(Save content first then metadata)